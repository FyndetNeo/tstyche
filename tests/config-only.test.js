import { afterEach, describe, expect, test } from "@jest/globals";
import { clearFixture, writeFixture } from "./__utils__/fixtureFactory.js";
import { normalizeOutput } from "./__utils__/normalizeOutput.js";
import { spawnTyche } from "./__utils__/spawnTyche.js";

const isStringTestText = `import { expect, test } from "tstyche";
test("is string?", () => {
  expect<string>().type.toBeString();
});
`;

const tsconfig = {
  extends: "../tsconfig.json",
  include: ["**/*"],
};

const fixture = "config-only";

afterEach(async () => {
  await clearFixture(fixture);
});

describe("'--only' command line option", () => {
  test("selects tests to run", async () => {
    const testText = `import { expect, test } from "tstyche";
test("external is string?", () => {
  expect<string>().type.toBeString();
});

test("external is number?", () => {
  expect<number>().type.toBeNumber();
});

test("internal is string?", () => {
  expect<string>().type.toBeString();
});
`;

    await writeFixture(fixture, {
      ["__typetests__/dummy.test.ts"]: testText,
      ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixture, ["--only", "external"]);

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(stderr).toBe("");

    expect(exitCode).toBe(0);
  });

  test("selects test group to run", async () => {
    const testText = `import { describe, expect, test } from "tstyche";
describe("external", () => {
  test("is string?", () => {
    expect<string>().type.toBeString();
  });
});

test("external is number?", () => {
  expect<number>().type.toBeNumber();
});

test("internal is string?", () => {
  expect<string>().type.toBeString();
});
`;

    await writeFixture(fixture, {
      ["__typetests__/dummy.test.ts"]: testText,
      ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixture, ["--only", "external"]);

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(stderr).toBe("");

    expect(exitCode).toBe(0);
  });

  test("does not override the '.skip' run mode flag", async () => {
    const testText = `import { expect, test } from "tstyche";
test("external is string?", () => {
  expect<string>().type.toBeString();
});

test.skip("external is number?", () => {
  expect<number>().type.toBeNumber();
});

test("internal is string?", () => {
  expect<string>().type.toBeString();
});
`;

    await writeFixture(fixture, {
      ["__typetests__/dummy.test.ts"]: testText,
      ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixture, ["--only", "external"]);

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(stderr).toBe("");

    expect(exitCode).toBe(0);
  });

  test("when '--skip' command line option is specified", async () => {
    const testText = `import { expect, test } from "tstyche";
test("external is string?", () => {
  expect<string>().type.toBeString();
});

test("external is number?", () => {
  expect<number>().type.toBeNumber();
});

test("internal is string?", () => {
  expect<string>().type.toBeString();
});
`;

    await writeFixture(fixture, {
      ["__typetests__/dummy.test.ts"]: testText,
      ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixture, ["--only", "external", "--skip", "number"]);

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(stderr).toBe("");

    expect(exitCode).toBe(0);
  });

  test("when search string is specified before the option", async () => {
    const testText = `import { expect, test } from "tstyche";
test("external is string?", () => {
  expect<string>().type.toBeString();
});

test("external is number?", () => {
  expect<number>().type.toBeNumber();
});

test("internal is string?", () => {
  expect<string>().type.toBeString();
});
`;

    await writeFixture(fixture, {
      ["__typetests__/dummy.test.ts"]: testText,
      ["__typetests__/isString.tst.ts"]: isStringTestText,
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixture, ["dummy", "--only", "external"]);

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(normalizeOutput(stderr)).toBe("");

    expect(exitCode).toBe(0);
  });

  test("when search string is specified after the option", async () => {
    const testText = `import { expect, test } from "tstyche";
test("external is string?", () => {
  expect<string>().type.toBeString();
});

test("external is number?", () => {
  expect<number>().type.toBeNumber();
});

test("internal is string?", () => {
  expect<string>().type.toBeString();
});
`;

    await writeFixture(fixture, {
      ["__typetests__/dummy.test.ts"]: testText,
      ["__typetests__/isString.tst.ts"]: isStringTestText,
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixture, ["--only", "external", "dummy"]);

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(normalizeOutput(stderr)).toBe("");

    expect(exitCode).toBe(0);
  });
});
