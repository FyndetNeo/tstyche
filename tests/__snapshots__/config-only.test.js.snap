// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`'--only' command line option does not override the '.skip' run mode flag: stdout 1`] = `
"uses TypeScript <<version>> with ./tsconfig.json

pass ./__typetests__/dummy.test.ts
  + external is string?
  - skip external is number?
  - skip internal is string?

Targets:    1 passed, 1 total
Test files: 1 passed, 1 total
Tests:      2 skipped, 1 passed, 3 total
Assertions: 2 skipped, 1 passed, 3 total
Duration:   <<timestamp>>

Ran tests matching 'external' in all test files.
"
`;

exports[`'--only' command line option selects test group to run: stdout 1`] = `
"uses TypeScript <<version>> with ./tsconfig.json

pass ./__typetests__/dummy.test.ts
  external
    + is string?
  + external is number?
  - skip internal is string?

Targets:    1 passed, 1 total
Test files: 1 passed, 1 total
Tests:      1 skipped, 2 passed, 3 total
Assertions: 1 skipped, 2 passed, 3 total
Duration:   <<timestamp>>

Ran tests matching 'external' in all test files.
"
`;

exports[`'--only' command line option selects tests to run: stdout 1`] = `
"uses TypeScript <<version>> with ./tsconfig.json

pass ./__typetests__/dummy.test.ts
  + external is string?
  + external is number?
  - skip internal is string?

Targets:    1 passed, 1 total
Test files: 1 passed, 1 total
Tests:      1 skipped, 2 passed, 3 total
Assertions: 1 skipped, 2 passed, 3 total
Duration:   <<timestamp>>

Ran tests matching 'external' in all test files.
"
`;

exports[`'--only' command line option when '--skip' command line option is specified: stdout 1`] = `
"uses TypeScript <<version>> with ./tsconfig.json

pass ./__typetests__/dummy.test.ts
  + external is string?
  - skip external is number?
  - skip internal is string?

Targets:    1 passed, 1 total
Test files: 1 passed, 1 total
Tests:      2 skipped, 1 passed, 3 total
Assertions: 2 skipped, 1 passed, 3 total
Duration:   <<timestamp>>

Ran tests matching 'external' and not matching 'number' in all test files.
"
`;

exports[`'--only' command line option when search string is specified after the option: stdout 1`] = `
"uses TypeScript <<version>>

pass ./__typetests__/dummy.test.ts
  + external is string?
  + external is number?
  - skip internal is string?

Targets:    1 passed, 1 total
Test files: 1 passed, 1 total
Tests:      1 skipped, 2 passed, 3 total
Assertions: 1 skipped, 2 passed, 3 total
Duration:   <<timestamp>>

Ran tests matching 'external' in test files matching 'dummy'.
"
`;

exports[`'--only' command line option when search string is specified before the option: stdout 1`] = `
"uses TypeScript <<version>>

pass ./__typetests__/dummy.test.ts
  + external is string?
  + external is number?
  - skip internal is string?

Targets:    1 passed, 1 total
Test files: 1 passed, 1 total
Tests:      1 skipped, 2 passed, 3 total
Assertions: 1 skipped, 2 passed, 3 total
Duration:   <<timestamp>>

Ran tests matching 'external' in test files matching 'dummy'.
"
`;
